contracts/Marketplace.sol

```diff
@@ -324,6 +327,10 @@ contract Marketplace is Ownable, IMarketplace2 {
         if (requirePayment){
             price = getPriceInData(addSeconds, p.pricePerSecond, p.priceCurrency);
             require(datacoin.transferFrom(msg.sender, p.beneficiary, price), "error_paymentFailed");
+            if(price > 0 && txFee > 0){
+                uint256 fee = txFee.mul(price).div(1 ether);
```

Comment:
So `txFee` is currently defined as a fraction with 18 decimals, correct? Is there a risk of overflow here in the multiplication? 18 decimals is way more precision than what's required, and the precision can easily be reduced if there's any risk of overflow. Basically 4 decimals would be enough (e.g. a 3.5% fee would be 0.035), or say 5-6 decimals to be on the safe overkill side.

Reply:
I don't think there's a risk of overflow because I'm using [SafeMath](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/math/SafeMath.sol) mul() which throws if there's a overflow error.

Re precision, I thought the same thing, but Juuso convinced me to use wei for the adminFee. I'm using it here for consistency.

Reply:
well, a "trick" if you're afraid of overflow, would be to `txFee.div(1 gwei).mul(price).div(1 gwei)`, but as @isentropy pointed out, the overflow would be noticed for sure. Also, the 32-byte `uint` range is something like 0...2^256 ~= 0...10^78, so you could multiply quite a few 10^18-numbers before hitting that.

Reply:
Copy-pasting from my message on slack for good measure:

- We know the valid range of txFee: it's max value is 100%, i.e. 1, i.e. 10^18.
2^256 is around 10^77. As 10^18 * 10^18 = 10^36, that leaves us around 10^41 of "full token" range.
- Assumption: individual purchases won't be worth more than a million USD, i.e. 10^6
- Assumption: DATA price won't go lower than 0.001 USD
- => purchases won't be more than 10^9 DATA, leaving us with 10^32 headroom which is plenty.

Resolving.